### ISSUE TITLE
gep type does not match ('LLVM3.8.0' Reader: 'LLVM 3.8.0')

### ISSUE THREAD

**Issue by BeanBagKing** [pocl/pocl#360]
I'm honestly not sure if this is a POCL issue, or an LLVM issue. It occurred while using POCL, so I believe it may be POCL using the incorrect name, and not LLVM. Please correct me if I'm wrong though and I'll close this issue and open one with LLVM.

The entire walkthrough of my installation and setup can be found here: http://www.nullsec.us/hashcat-3-0-on-a-32-bit-linux-vm/

The exact error message is below
`error: Explicit gep type does not match pointee type of pointer operand (Producer: 'LLVM3.8.0' Reader: 'LLVM 3.8.0')`

Basically, install dependencies and build POCL from git. Use POCL's OpenCL libraries to run hashcat with --force -b, and about 3/4ths of the way down, it seems to try to call LLVM 3.8 by the wrong name (note the space).

If there's any other logs that would be helpful, I'll try to provide them, but I'm not sure what else to give here.

**Comment by pjaaskel** [comment]
I've seen this before and it was mentioned by @dsandersimgtec in the email list, but I cannot remember what made it go away. Should not be a pocl issue. Perhaps you have a bitcode kernel builtin lib built with another revision of LLVM with that slight white space difference in the version string?

**Comment by dsandersimgtec** [comment]
The one I saw was a different error (corrupted bitcode). I'm fairly certain that the whitespace between LLVM and 3.8.0 is misleading since the check that causes LLVM to emit the '(Producer: ... Reader: ...)' bit is only checking for an non-empty producer string. It doesn't actually compare the two.

I can't tell from the error whether the cause is in clang, llvm, or pocl. Are pocl's optimization passes manadatory? If not, then I'd suggest skipping them and seeing if you still get the error. If it goes away then it's probably one of pocls optimizations but if it still happens then it's probably clang/llvm.

**Comment by BeanBagKing** [comment]
> Are pocl's optimization passes mandatory? If not, then I'd suggest skipping them and seeing if you still get the error.

I have no idea if they are mandatory, or how to skip them if they aren't.:(

**Comment by pjaaskel** [comment]
For CPUs they are. If you can isolate a piece of kernel that produces it, it would help.

**Comment by BeanBagKing** [comment]
I don't think I'll be able to, but if I ever get to that point, I'll be sure to return. In the mean time I think the best I can offer is running any commands you want and letting you know what the output is (in other words, I'm over my head and have no idea what I'm doing, sorry).

**Comment by pjaaskel** [comment]
Try removing .cache/pocl in your home. Maybe a corrupted kcache.

**Comment by pavanky** [comment]
@pjaaskel We are seeing the same bug in arrayfire (https://github.com/arrayfire/arrayfire/issues/1553).

Deleting the cache as you suggested fixes it for run, but subsequent runs are still buggy. Deleting the cache fixes it for one run again.

The OpenCL kernel that reproduces this issue in arrayfire can be seen here: 
https://github.com/arrayfire/arrayfire/blob/devel/src/backend/opencl/kernel/random.cl

The kernel is compiled with the following options:
```
 -cl-std=CL2.0 -D dim_t=long -D T=float -D repeat=32 -D randu
```

**Comment by pavanky** [comment]
FYI, running `llvm-dis` on the cached bitcode file causes the same issue. Could it be a potential bug in llvm 3.8 ?

**Comment by pjaaskel** [comment]
Can (any of) you reproduce this now with LLVM 3.9?

**Comment by pjaaskel** [comment]
Assuming transient bug until there's a way to reproduce it.
